wgpu/src/lib.rs

```diff
@@ -1465,6 +1466,12 @@ pub struct VertexState<'a> {
     /// The name of the entry point in the compiled shader. There must be a function with this name
     /// in the shader.
     pub entry_point: &'a str,
+    /// Specifies the values of pipeline-overridable constants in the shader module module.
+    ///
+    /// The key represents the numeric ID of the constant, if one is specified, and otherwise the constant's identifier name.
```

Comment:
> The key represents the numeric ID of the constant, if one is specified, and otherwise the constant's identifier name.

I don't think I understand this. Does this mean that the key might be one of (1) `String` containing ASCII numeric digits _or_ (2) a typical C-style identifier like might be written in a shader?

Doing further reading in ``7.2.2 `override` Declaration`` from [the spec.](https://gpuweb.github.io/gpuweb/wgsl/#override-decls), it seems that option (1) corresponds to the `id(…)` attribute:

> If the declaration has an [id](https://gpuweb.github.io/gpuweb/wgsl/#attribute-id) attribute applied, the literal operand is known as the pipeline constant ID.
>
> … The constant is identified by a pipeline-overridable constant identifier string[](https://gpuweb.github.io/gpuweb/wgsl/#pipeline-overridable-constant-identifier-string), which is the base-10 representation of the [pipeline constant ID](https://gpuweb.github.io/gpuweb/wgsl/#pipeline-constant-id) if specified, and otherwise the declared [name](https://gpuweb.github.io/gpuweb/wgsl/#name) of the constant.

However, that seems like `String` would be the wrong type for this, since we'd have to parse the ID as a `u16` anyway for validation, right? Am I missing something?

---

This question applies to other places where this doc has been copy-pasted.

Reply:
> if specified, and otherwise the declared [name](https://gpuweb.github.io/gpuweb/wgsl/#name) of the constant.

They key must be the [pipeline constant ID](https://gpuweb.github.io/gpuweb/wgsl/#pipeline-constant-id) (`u16`) if one was specified via the `@id()` attribute or the [name](https://gpuweb.github.io/gpuweb/wgsl/#name)/[identifier](https://gpuweb.github.io/gpuweb/wgsl/#identifier) of the pipeline-overridable constant.

Since the `key` is a `String` in the IDL, we keep it as such until we reach naga where we will validate the necessary invariants.

If `wgpu` users want more type expressiveness for the `constants` we can update its API later to accommodate that. It could look like:

```rust
enum Key {
  Id(u16),
  Name(&str)
}
enum Value {
  F32(f32),
  U32(u32),
  I32(i32),
  Bool(bool),
}
constants: &HashMap<Key, Value> or &[(Key, Value)]
```

cc @cwfitzgerald

Reply:
I'm not particularly happy with the untyped api. At the very least, I think we should add the Key type. I'm not sure there's actually any benefit from strongly typing the Value here.

Reply:
`Value` could implement `From` for all the types which would make it easier to use. The benefit being that you wouldn't have to allocate `String`s and convert to and from them unnecessarily.

Reply:
What is your take on?

```rust
constants: &'a HashMap<Key, Value> or &'a [(Key<'a>, Value)]
```

Note that `Key::Name` would have to be a `String` for the `HashMap` but could be a `&str` for the slice of tuples.

Reply:
I'm definitely leaning array - hash maps are hard to construct inline and we don't really need deduplication behavior at the edge.

Reply:
Thinking a bit more about this, I think we should come up with a solution that works for all users of this API (Firefox, deno and wgpu) since all have to pass this structure through `wgpu-core` -> `wgpu-hal` -> `naga`.

Reply:
I'll take a stab at trying to update it to the shape below and report back.

```rust
enum Key<'a> {
  Id(u16),
  Name(&'a str)
}
constants: &'a [(Key<'a>, f64)]
```

Reply:
Yeah, we can leave it for now and iterate on it once it's in.